### PR TITLE
Prevent perpetual use of  expired auth signature; remove superfluous API functions

### LINKED ISSUE #324: Expose `registerEncrypters` from `taco` API

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**

- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [x] Other

**Required reviews:**

- [ ] 1
- [ ] 2
- [x] 3

**What this does:**

> High-level idea of the changes introduced in this PR. List relevant API
> changes (if any), as well as related PRs and issues.

Based over #560 (somewhat of a follow-up PR).

- `EIP4361AuthProvider` caches the auth signature it first generates perpetually (unless storage is explicitly cleared eg. `localStorage.clear()` in apps). However, this auth signature can become expired (> 2h old), and so the provider should detect that on cache retrieval and regenerate the auth signature if it becomes expired.
- Add test for `SingleSignOnEIP4361AuthProvider`
- Clean up tests so that signer is obtained from provider, instead of creating separate `fakeProvider()` and `fakeSigner()`; this way they are better linked for tests. This only applies to tests - in real life it is possible to have a signer with no provider, but this was not the case for tests prior to this change anyway.
- Remove no longer applicable/superfluous API calls from being exposed:
   - `isAuthorized(...)` - calls into the `Coordinator` contract to call this function, but we are moving away from this where calls should be made directly to the `AccessController`. However the accessController only checks for `isEncryptionAuthorized` and doesn't provide the ability to check if an encrypted address is authorized - which is what is implemented here. Therefore, it only applies to the GlobalAllowList contracts, but it's unclear if we need that ability in `taco-web`
   - `registerEncrypters(...)` - Was never publicly exposed and it is unclear if we need this ability in `taco-web`. #324 alluded to it, but not sure if we still care. Let me know if we do (cc @arjunhassard ).
   
   *If we decide down the road we need them, we can add them back.*

**Issues fixed/closed:**

> - Fixes #...

Closes https://github.com/nucypher/taco-web/issues/324

**Why it's needed:**

> Explain how this PR fits in the greater context of the NuCypher Network. E.g.,
> if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

> What should reviewers focus on? Is there a particular commit/function/section
> of your PR that requires more attention from reviewers?

### CODE REVIEW

File: examples/taco/nodejs/src/index.ts [hunk 1]
```diff
@@ -81,18 +80,6 @@ const encryptToBytes = async (messageString: string) => {
     encryptorSigner,
   );
```
Comment (manumonti) [inline examples/taco/nodejs/src/index.ts:82]: Are we still "checking contract compatibility" even when removing this from the Testnet examples that are run every hour?

https://github.com/nucypher/taco-web/actions/workflows/lynx.yml

**Review by manumonti — APPROVED** [review]
LGTM! I've just left a concern I have.

File: examples/taco/nodejs/src/index.ts (hunk 1 above)
Comment (derekpierre) [reply examples/taco/nodejs/src/index.ts:82]: At the moment, I don't think `isAuthorized` is useful as part of the exposed `taco` API, hence the removal here. If others feel that it is useful I can put it back.

It's unclear to me how comprehensive this check is as it pertains to "checking compatibility" based on one function.  Do you have any additional insight there, perhaps I'm missing something?  Perhaps testing the `DkgCoordinatorAgent` with `lynx` contracts could be a better solution to CI checking compatibility...maybe - not sure what that entails or if it is feasible?

Additionally, `Coordinator.isEncryptionAuthorized()` which is called by `isAuthorized()`, is actually deprecated in the latest Coordinator contract, and will be removed down the road. Instead, the underlying acccessController for the ritual Id should be called directly for that check.

Comment (manumonti) [reply examples/taco/nodejs/src/index.ts:82]: No, I don't have a clear idea about this check beyond that we can check that this specific method on the Coordinator contract has not been modified. 

So yes, I agree we can delete this.

My concern came after I read the commented lines in the code talking about "checking contract compatibility": I was wondering if I'm missing something here.

**Review by vzotova — APPROVED** [review]

File: packages/taco-auth/src/providers/eip4361/eip4361.ts [hunk 2]
```diff
@@ -61,6 +67,19 @@ export class EIP4361AuthProvider {
     return authMessage;
   }
 
+  private isMessageExpired(message: string): boolean {
+    const siweMessage = new SiweMessage(message);
+    if (!siweMessage.issuedAt) {
+      // don't expect to ever happen; but just in case
+      return false;
+    }
```
Comment (KPrasch) [inline packages/taco-auth/src/providers/eip4361/eip4361.ts:75]: This is a sentinel against `undefined`?

**Review by KPrasch — APPROVED** [review]
Overall looks good 👍🏻 While I do find this acceptable (and very convenient for implementers) it's interesting to observe the lines between application and library logic blurring a bit here.

File: packages/taco-auth/src/providers/eip4361/eip4361.ts (hunk 2 above)
Comment (derekpierre) [reply packages/taco-auth/src/providers/eip4361/eip4361.ts:75]: Yes.

From the EIP4361 spec,`issued-at` is a required value - https://docs.login.xyz/general-information/siwe-overview/eip-4361#specification.

However, in the `siwe` typescript library implementation, it is an optional value in the constructor - https://github.com/spruceid/siwe/blob/main/packages/siwe/lib/client.ts#L49. As a result, Typescript complains if we don't check whether it is defined first before using it.

**Comment by derekpierre** [comment]
> Overall looks good 👍🏻 While I do find this acceptable (and very convenient for implementers) it's interesting to observe the lines between application and library logic blurring a bit here.

Because we opted to cache the message ourselves to prevent users having to sign every time for proof of wallet ownership, it forces us to manage the cache appropriately - this PR does the latter. Definitely open to ideas here if you have any.

File: packages/taco-auth/src/providers/eip4361/eip4361.ts (hunk 2 above)
Comment (cygnusv) [reply packages/taco-auth/src/providers/eip4361/eip4361.ts:75]: Is this a divergence of `siwe` wrt to the spec then?

**Review by cygnusv — APPROVED** [review]

File: packages/taco-auth/src/providers/eip4361/eip4361.ts (hunk 2 above)
Comment (derekpierre) [reply packages/taco-auth/src/providers/eip4361/eip4361.ts:75]: As far as I can tell, yes, it is a divergence.

Comment (derekpierre) [reply packages/taco-auth/src/providers/eip4361/eip4361.ts:75]: Well, it's not the constructor but the SiweMessage struct - so still a deviation - https://github.com/spruceid/siwe/blob/main/packages/siwe/lib/client.ts#L49

Without the check, the compiler returns the following:
```
packages/taco-auth/src/providers/eip4361/eip4361.ts:72:36 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(value: string | number | Date): Date', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type 'string | number | Date'.
      Type 'undefined' is not assignable to type 'string | number | Date'.
  Overload 2 of 4, '(value: string | number): Date', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type 'string | number'.
      Type 'undefined' is not assignable to type 'string | number'.
 ```
 
 I'll file an issue for the `siwe` library, and perhaps they can either clarify or fix.

**Pull request merged**